### ISSUE TITLE
deprecateTypingAliases incorrectly reports Optional as deprecated

### ISSUE THREAD

**Issue by jpgoldberg** [microsoft/pyright#9793]
## Environment data

-   Pylance version: 2024.12.1
-   OS and version: macOS 15.2
-   Python version 3.13.1 (installed with pyenv)
-  Visual Studio Code: Version: 1.96.4 (Universal)

## Code Snippet

An example using `typing.Optional`"

```python
from typing import Optional


def wrap_division(dividend: float, divisor: float) -> Optional[float]:
    """A silly way to wrap division, but it illustrates Optional"""
    try:
        return dividend / divisor
    except ZeroDivisionError:
        return None
```

`.vscode/settings.json`
```json
{
    "spellright.language": [
        "en"
    ],
    "spellright.documentTypes": [
        "markdown",
        "latex",
        "plaintext"
    ],
    "python.analysis.typeEvaluation.deprecateTypingAliases": true
}
```

## Repro Steps

1. Set Python Interpreter in VSCode to 3.13.1 (others may show similar, but this is where I encountered this)
2. Turn on `deprecateTypingAliases`
3. import and use `Optional` from `typing`

## Expected behavior

No deprecation warnings for Optional.

## Actual behavior

`Optional` is marked (struck through) as deprecated with a message from Pylance saying,

> This type is deprecated as of Python 3.10; use "| None" instead

![Screenshot with deprecation warning](https://github.com/user-attachments/assets/9d1779c9-eed1-41b3-83d6-31b2ccba0ddd)

### Why this is unexpected

No Python documentation or approved PEPs that I have been able to find lists `Optional` as deprecated. Although `X | None` is an alternative to `Optional[X]` there is no indication in standard library documentation that the former is even preferred, much less that the latter is deprecated.

The [Deprecated aliases section](https://docs.python.org/3/library/typing.html#deprecated-aliases) of the typing module docs for the current version (3.13.1) does not list `Optional` as deprecated,
and the [documentation for Optional](https://docs.python.org/3/library/typing.html#typing.Optional) says nothing about deprecation. It does say,

>  Changed in version 3.10: Optional can now be written as X | None. See [union type expressions](https://docs.python.org/3/library/stdtypes.html#types-union).

but that is not a deprecation statement. It even says "can" instead of "should".

My guess is that the incorrect treatment of this arises from a misreading of [PEP 585](https://peps.python.org/pep-0585/), which does deprecate a number of things from the `typing` module. But no where does it suggest that `Optional` is deprecated.

**Comment by erictraut** [comment]
This behavior is intended, not a bug. The `typing.Optional` symbol is no longer recommended. While it may not be formally deprecated, it is generally recommended that you use the shorter and easier-to-read `X | None`. If you're interested in modernizing your code base, you should switch to `X | None`. This rule is intended to help pyright users identify such usage. If you want to continue using `Optional` in your code base, you can ignore the strike-through and/or leave the `deprecateTypingAliases` option set to false, which is its default value.

**Comment by jpgoldberg** [comment]
Thank you. Although I disagree with your design choice, I now understand that it is the result of deliberate decision that took into account the fact that Optional is not officially deprecated.  And I acknowledge that it is the job of linters to be opinionated, even when I may disagree with some of those opinions.

I do want to see when I use typing aliases that are truly deprecated, which is why I enabled `deprecateTypingAliases` in the first place, but I will switch that off and rely on other linting tools to catch those.

Again, thank you for your response and for your work on Pylance.

**Comment by smuuf** [comment]
`Optional` is NOT deprecated, period. It is not a correct behavior for Pyright to say it is.

**Comment by ShaneMurphy2** [comment]
> `Optional` is NOT deprecated, period. It is not a correct behavior for Pyright to say it is (and to treat it as such).

Agreed, this should be under a rule called "reportNonStandardTypingStyle" or something.

**Comment by nemahs** [comment]
> This behavior is intended, not a bug. The `typing.Optional` symbol is no longer recommended. While it may not be formally deprecated, it is generally recommended that you use the shorter and easier-to-read `X | None`. If you're interested in modernizing your code base, you should switch to `X | None`. This rule is intended to help pyright users identify such usage. If you want to continue using `Optional` in your code base, you can ignore the strike-through and/or leave the `deprecateTypingAliases` option set to false, which is its default value.

Can you point to documentation that says that Optional is no longer recommended, because as others have mentioned it's nowhere I can find in the PEPs or the Python documentation.

**Comment by jpgoldberg** [comment]
The more I think about this, the more I consider it a bug.  Please reopen the issue, @erictraut or @rchiodo 

You may wish to discourage some behavior with your linting, but **it is an error to tell users that something is deprecated when it is not deprecated**.  Among other things, it can lead users to search in vain for an official deprecation announcement, and it can lead them to pass on incorrect information.

Find some other wording to discourage certain non-deprecated constructs you wish to discourage.

**Comment by erictraut** [comment]
> The more I think about this, the more I consider it a bug.

This isn't a bug. It's a very intentional choice. 

None of these outdated special forms (`List`, `Dict`, `Tuple`, `Union`, `Optional`) are likely to ever be removed from the `typing` module because of the large amount of old code that relies upon them. So if you want to continue to use them in your code, you certainly can. For new code, I recommend avoiding them because 1) they require extra (unnecessary) imports, 2) are confusing and more difficult to read for developers who are new to type annotations, and 3) make your code look outdated. 

The typing community has been working over the past several years to eliminate the need for these outdated constructs and encouraging users to move away from them. For example, you won't see any of these outdated special forms used in the community-maintained typeshed library.

You can enable pyright's `deprecateTypingAliases` check to avoid using these outdated constructs in new code. For existing code, you can keep `deprecateTypingAliases` disabled. It is off by default, even in strict type checking mode, so it's completely up to you if you'd like to enable it.

If you'd prefer to apply finer-grained rules about which of these outdated forms you want to accept within your code base, you can use a linter like `ruff`. It provides rules like [this one](https://github.com/astral-sh/ruff/issues/4858).

**Comment by smuuf** [comment]
> None of these outdated special forms (List, Dict, Tuple, Union, Optional, etc.) are likely to ever be removed...

You list these as having equivalent status, but the fact is that `List`, `Dict`, `Tuple` _(and `Set`, `FrozenSet`, `Type`, `OrderedDict`, etc.)_ _really are_ deprecated.

And both `Union` and `Optional` _are not_ deprecated - and not even considered outdated - as seen here: https://docs.python.org/3/library/typing.html#aliases-to-built-in-types _(It's somehow ironic that I need to send this link and base my argument on the official Python docs to show that Pyright's behavior is incorrect.)_.

I do want Pyright to inform me about usage of _deprecated_ type aliases like `List[...]` or `Dict[...]`, so disabling `deprecateTypingAliases` isn’t the right solution. But I’d also like Pyright not to tell me that `Optional[bool]` is deprecated - because it isn't. 🤷 

For me `Optional[Something]` ("optionally Something") carries different semantics than `Something | None` ("Something or None"). I want to express that kind of intent in my function parameter types. But Pyright tells me it’s wrong (deprecated). But it's not (deprecated). ☹

**Comment by d1b** [comment]
Is there any updates with regards to this issue? 

Is the fix mostly just - 
```diff --git a/packages/pyright-internal/src/analyzer/deprecatedSymbols.ts b/packages/pyright-internal/src/analyzer/deprecatedSymbols.ts
index b5c65c938..4f475f5e9 100644
--- a/packages/pyright-internal/src/analyzer/deprecatedSymbols.ts
+++ b/packages/pyright-internal/src/analyzer/deprecatedSymbols.ts
@@ -301,8 +301,6 @@ export const deprecatedAliases = new Map<string, DeprecatedForm>([
 ]);
 
 export const deprecatedSpecialForms = new Map<string, DeprecatedForm>([
-    ['Optional', { version: pythonVersion3_10, fullName: 'typing.Optional', replacementText: '| None' }],
-    ['Union', { version: pythonVersion3_10, fullName: 'typing.Union', replacementText: '|' }],
     [
         'Callable',
         {
```

or should it be enabled/disabled by a specific config option perhaps?

**Comment by leycec** [comment]
**This issue triggers @beartype maintainer @leycec.** The PEP 484-compliant `typing.Union[...]` and `typing.Optional[...]` type hint factories are both self-caching. The PEP 604-compliant `|` operator is *not*. This has profound space complexity consequences for runtime:

```python
# Import boring machinery. So boring, yet it actually caches. So, not boring at all.
>>> from typing import Optional

# Cache everything, "typing" module! Cache it so hard.
>>> Optional[int] is Optional[int]
True  # <-- *GOOD*. This is good.

# Cache nothing, "|" operator! Why are you so sucky?
>>> int | None is int | None
False  # <-- *UH OH*. We are now panicking.
```

The same also holds true for genuinely deprecated PEP 484-compliant type hint factories like `typing.List[...]`. In other words, all of those things that PEP 585 deprecated? Uh. Yeah. They shouldn't have been. The syntax may be a bit clumsier, but those things that CPython devs were so hasty to deprecate were actually functionally superior to their PEP 585-compliant analogues.

`typing.Union[...]` and `typing.Optional[...]` are no different. They self-cache, too. Except they are different. They're not deprecated. By any reasonable reading of reality, they're superior to PEP 604-compliant `|`-delimited unions. Yet again, the syntax is clumsier. And? What of it? Since when is useless syntactic sugar of greater concern to anyone over runtime semantics that actually reduce space consumption?

**Pyright chose poorly here.**

![](https://media1.tenor.com/m/_UvSPt2NmzIAAAAd/ares-god-of-war.gif)

**Comment by Ige-kun** [comment]
> there is no indication in standard library documentation that the former is even preferred

> Can you point to documentation that says that Optional is no longer recommended, because as others have mentioned it's nowhere I can find in the PEPs or the Python documentation.

Since Python 3.10, [the documentation has stated that the shorthand syntax is "recommended" over the `typing` imports.](https://docs.python.org/3.10/library/typing.html#typing.Union:~:text=the%20shorthand%20int%20%7C%20str%2E%20Using%20that%20shorthand%20is%20recommended)
